design/open/hard-delete.md

```diff
@@ -0,0 +1,400 @@
+      2. Else - add branch staging token to reference list and perform `set if`
+         1. If predicate failed - retry
+   2. If key not found - assume address was committed - perform copy without adding a new reference
+
+#### Commit
+
+1. Perform commit flow in the same manner as today
+2. On successful commit - create entry using `commit_id` and ordered list of tokens under the `committed` prefix 
+
+#### Reset / Delete Branch
+
+1. Perform reset / delete flow in the same manner as today
+2. On successful execution - create entries for all staging tokens under the `deleted` prefix
+
+### Background delete job
+
+* Scan committed tokens
+    * For each committed object - remove reference key
+    * For each staging token in committed entry - "Move staging token to deleted"   
+      All objects in committed staging tokens that were not actually committed are candidates for deletion, therefore
+  we can either execute the deleted tokens logic, or create entries for these tokens under the deleted prefix (implementation detail)
+    * Remove `staging/committed` key
+    * Delete staging area
+* Scan deleted tokens  
+  * For each token - remove references for all objects on staging token
+  * If it is the last reference for that object (list is empty after update) - perform hard-delete
+  * Delete staging area
+
+### Handling key override on same staging token - improvement
+
+TBD
+
+## 5. No Copies
+
+The idea is to make sure we do not enable copies, when no two entries in stage will point to the same physical object we will enable delete of the physical address in the following:
+
+- Upload - get previous entry and delete previous physical address unless staging token was updated
+- Revert - branch / prefix / object, when entry is dropped we can delete the physical address
+- Delete - repo/branch/object, delete the physical address after entry no longer exists
+
+We should enable the following to enable the physical delete:
+
+- Make sure we don't have the same physical address by sign links we return in our API. The logical path will be part of the signature that will be part of the physical address we generate. This will enable us to block and use of a physical address outside the repo/branch/entry.
+- The S3 gateway `put` operation with the copy support will use the underlying adapter to copy the data. Will require from each adapter to support/emulate a 'copy' operation.
+* (optional) Enable 'move' API as alternative to 'copy' we have seen that moving objects from one location to another by copy+delete of metadata will enable easy support for lakeFSFS move. This can be implemented by marking the entry on staging as 'locked'. During the move operation - lock+copy+delete+unlock operation. In case of any failure, as we don't have transactions, we may keep two entries, or keep locked object without ever delete its physical address.
+
+TODO(Barak): list more cases and how it will be address in this solution.
+
+### Flows
+
+#### Move operation
```

Comment:
If it's optional can we start without?

Reply:
AFAIU this is a requirement. The move operation was created to support a main flow in our Spark client which currently is being performed on the metadata level (by using copy + delete).
As per @arielshaqed, this is something that we do not want to support by a full copy due to the performance implications

Reply:
Note that Spark lakeFSFS and Hadoop S3A do not exactly require a "move".  We currently use LinkPhysicalAddress followed by a delete.  So it isn't atomic, but at least it requires only metadata operations.